### ISSUE TITLE
Add assert_array_shape calls to method code

### ISSUE THREAD

**Issue by tomwhite** [pystatgen/sgkit#267]
The `assert_array_shape` call is used a lot in REGENIE, and serves both as a runtime check and as documentation for the reader. We should sprinkle it liberally through other functions.

**Comment by ravwojdyla** [comment]
Quoting from https://github.com/pystatgen/sgkit/pull/314#discussion_r502548783:
> I meant to ask this question many times actually, so we use these `assert`s pretty generously, which I like, but I wonder if we should ever worry about those assert statements getting [optimised out](https://docs.python.org/3/reference/simple_stmts.html#grammar-token-assert-stmt)?

**Comment by tomwhite** [comment]
Good point. I think we should perhaps document it, but I don't think we should come up with a new assertion mechanism.

**Comment by ravwojdyla** [comment]
@tomwhite yea, I suppose it could be part of this issue to add documentation.
